File: BinaryTreeLevelOrderTraversalII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> lists = new ArrayList<List<Integer>>();
        List<Integer> list = null;
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        if (root != null)
            queue.offer(root);
        
        while (!queue.isEmpty()) {
            list = new ArrayList<Integer>();
            int size = queue.size();
            
            while (size-- > 0) {
                TreeNode cur = queue.poll();
                list.add(cur.val);
                
                if (cur.left != null)
                    queue.offer(cur.left);
                if (cur.right != null)
                    queue.offer(cur.right);
            }
            lists.add(list);
        }
        
        Collections.reverse(lists);
        return lists;
    }
}

File: lulu/binaryTreeLevelOrderTraversalII.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.ans = []
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        self.helper(root, 0)
        ans = []
        for i in range(len(self.ans) - 1, -1, -1):
            ans.append(self.ans[i])
        return ans
    
    def helper(self, root, level):
        if root:
            if level >= len(self.ans):
                self.ans.append([])
                
            self.ans[level].append(root.val)
            self.helper(root.left, level + 1)
            self.helper(root.right, level + 1)
